File: sdks/python/src/memoturn/cohere.py
"""Drop-in wrapper for a Cohere client (``cohere`` v5+) — records ``client.chat`` and
``client.chat_stream`` as generations, handling both API generations in one wrapper:

- **ClientV2** (``cohere.ClientV2``): ``chat()`` returns ``{message: {role, content:
  [{type: "text", text}, ...], tool_calls}, usage: {tokens: {input_tokens, output_tokens},
  billed_units: {...}}}``; ``chat_stream()`` yields typed events discriminated by ``.type``
  (``content-delta`` carries ``event.delta.message.content.text``; ``message-end`` carries
  ``event.delta.usage``).
- **legacy Client** (v1 API): ``chat()`` returns ``{text, meta: {tokens: {input_tokens,
  output_tokens}}}``; ``chat_stream()`` yields events discriminated by ``.event_type``
  (``text-generation`` carries ``.text``; ``stream-end`` carries the full ``.response``).

Both shapes are probed defensively per response, so the same ``wrap_cohere`` works on
either client. Cohere reports token counts as floats and never a total — usage is
int-coerced and ``totalTokens`` computed as input + output."""
from __future__ import annotations

from typing import Any, Optional

from ._stream import _ABANDONED, _RecordingStream, _get
from .client import Memoturn, Trace
from .decorator import get_client

_EXCLUDED_PARAMS = ("model", "messages", "message", "chat_history")


def wrap_cohere(client: Any, memoturn: Optional[Memoturn] = None, *, trace: Optional[Trace] = None) -> Any:
    mt = memoturn or get_client()
    original_chat = client.chat
    original_stream = getattr(client, "chat_stream", None)

    def chat(*args: Any, **kwargs: Any) -> Any:
        gen = _start(mt, trace, kwargs)
        try:
            resp = original_chat(*args, **kwargs)
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise
        gen.end(output=_output(resp), usage=_usage_from_response(resp))
        return resp

    def chat_stream(*args: Any, **kwargs: Any) -> Any:
        gen = _start(mt, trace, kwargs)
        try:
            resp = original_stream(*args, **kwargs)
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise

        acc = _CohereStreamAccumulator()

        def on_done(err: Any) -> None:
            output, usage = acc.finalize()
            if err is _ABANDONED:
                gen.end(level="WARNING", statusMessage="stream ended before completion", output=output, usage=usage)
            elif err is not None:
                gen.end(level="ERROR", statusMessage=str(err), output=output, usage=usage)
            else:
                gen.end(output=output, usage=usage)

        return _RecordingStream(resp, acc.add, on_done)

    client.chat = chat  # type: ignore[assignment]
    if original_stream is not None:
        client.chat_stream = chat_stream  # type: ignore[assignment]
    return client


def _start(mt: Memoturn, trace: Optional[Trace], kwargs: dict) -> Any:
    t = trace or mt.trace(name="cohere.chat")
    model_parameters = {k: v for k, v in kwargs.items() if k not in _EXCLUDED_PARAMS}
    return t.generation(
        name="cohere.chat", model=kwargs.get("model"), provider="cohere",
        input=_input(kwargs), modelParameters=model_parameters,
    )


def _input(kwargs: dict) -> Any:
    if kwargs.get("messages") is not None:  # v2
        return kwargs["messages"]
    if kwargs.get("chat_history") is not None:  # v1 with history
        return {"chatHistory": kwargs["chat_history"], "message": kwargs.get("message")}
    return kwargs.get("message")  # v1


def _output(resp: Any) -> Any:
    try:
        message = _get(resp, "message")
        if message is not None:  # v2
            if hasattr(message, "model_dump"):
                return message.model_dump()
            return {"role": _get(message, "role") or "assistant", "content": _content_text(_get(message, "content"))}
        text = _get(resp, "text")
        if text is not None:  # v1
            return {"role": "assistant", "content": text}
        return resp
    except Exception:  # noqa: BLE001
        return resp


def _content_text(content: Any) -> str:
    """v2 message content is a list of typed items; concatenate the text ones."""
    if isinstance(content, str):
        return content
    parts = []
    for item in content or []:
        text = _get(item, "text")
        if isinstance(text, str):
            parts.append(text)
    return "".join(parts)


def _usage_from_response(resp: Any) -> Optional[dict]:
    usage = _get(resp, "usage")  # v2: {tokens, billed_units}
    if usage is not None:
        return _map_usage(usage)
    meta = _get(resp, "meta")  # v1: meta.tokens / meta.billed_units — same field names
    if meta is not None:
        return _map_usage(meta)
    return None


def _map_usage(container: Any) -> Optional[dict]:
    tokens = _get(container, "tokens")
    if tokens is None:
        tokens = _get(container, "billed_units")
    if tokens is None:
        return None
    prompt = _int(_get(tokens, "input_tokens"))
    completion = _int(_get(tokens, "output_tokens"))
    if prompt is None and completion is None:
        return None
    # Omit absent fields entirely — the ingest schema treats them as optional but
    # rejects explicit nulls; Cohere never reports a total, so it's computed.
    usage = {"promptTokens": prompt, "completionTokens": completion}
    if prompt is not None and completion is not None:
        usage["totalTokens"] = prompt + completion
    return {k: v for k, v in usage.items() if v is not None}


def _int(value: Any) -> Optional[int]:
    """Cohere reports token counts as floats; the wire contract wants nonnegative ints."""
    try:
        return int(value) if value is not None else None
    except (TypeError, ValueError):
        return None


class _CohereStreamAccumulator:
    """Accumulates text deltas across both stream generations (v2 ``.type`` events and v1
    ``.event_type`` events) and captures usage from the final event of either shape."""

    def __init__(self) -> None:
        self._text: list[str] = []
        self._usage: Optional[dict] = None

    def add(self, event: Any) -> None:
        kind = _get(event, "type") or _get(event, "event_type")
        if kind == "content-delta":  # v2 text delta
            delta = _get(event, "delta")
            message = _get(delta, "message")
            text = _get(_get(message, "content"), "text")
            if isinstance(text, str):
                self._text.append(text)
        elif kind == "text-generation":  # v1 text delta
            text = _get(event, "text")
            if isinstance(text, str):
                self._text.append(text)
        elif kind == "message-end":  # v2 final event
            usage = _map_usage(_get(_get(event, "delta"), "usage"))
            if usage is not None:
                self._usage = usage
        elif kind == "stream-end":  # v1 final event: carries the whole non-streamed response
            response = _get(event, "response")
            usage = _usage_from_response(response) if response is not None else None
            if usage is not None:
                self._usage = usage
            if not self._text:
                text = _get(response, "text")
                if isinstance(text, str):
                    self._text.append(text)

    def finalize(self) -> tuple[Any, Optional[dict]]:
        output = {"role": "assistant", "content": "".join(self._text)} if self._text else None
        return output, self._usage

File: sdks/js/src/cohere.ts
import type { Memoturn, MemoturnTrace } from "./client.js";
import { tapStream } from "./stream.js";

/**
 * Drop-in wrapper for a Cohere client (`cohere-ai` v7+) — records the chat surface as memoturn
 * generations (model, params, usage, latency, output), across both API generations:
 *
 * - **v1** (`CohereClient`): `client.chat(...)` / `client.chatStream(...)` — response is
 *   `{ text, meta: { tokens } }`, streamed events carry `eventType` (`"text-generation"` deltas,
 *   final `"stream-end"` with the full response).
 * - **v2** (`CohereClientV2`, or the `client.v2` namespace on a v1 client): `chat(...)` /
 *   `chatStream(...)` — response is `{ message: { content: [...] }, usage: { tokens } }`,
 *   streamed events carry `type` (`"content-delta"` deltas, final `"message-end"` with usage).
 *
 * Because `CohereClientV2` serves v2 shapes from the same top-level `chat`/`chatStream` method
 * names, the wrapper sniffs the response/event shape rather than assuming one per method — both
 * client classes and the `client.v2` namespace are covered by the same interception. Cohere
 * reports `inputTokens`/`outputTokens` with no total; `totalTokens` is computed as their sum
 * when both are present.
 *
 *   const cohere = wrapCohere(new CohereClientV2({ token }), memoturn);
 *   await cohere.chat({ model: "command-a-03-2025", messages });
 *
 * Pass `{ trace }` to nest generations under an existing trace; otherwise each call gets its
 * own trace. Pass `{ streamTimeoutMs }` to override the idle-stream abandonment backstop
 * (default 120s).
 */

type WrapOptions = { trace?: MemoturnTrace; traceName?: string; streamTimeoutMs?: number };

/** Cohere reports `inputTokens`/`outputTokens` (no total) — map and sum when both present. */
function mapUsage(tokens: any): Record<string, number> | undefined {
  if (!tokens) return undefined;
  const usage: Record<string, number> = {};
  if (typeof tokens.inputTokens === "number") usage.promptTokens = tokens.inputTokens;
  if (typeof tokens.outputTokens === "number") usage.completionTokens = tokens.outputTokens;
  if (typeof tokens.inputTokens === "number" && typeof tokens.outputTokens === "number") {
    usage.totalTokens = tokens.inputTokens + tokens.outputTokens;
  }
  return Object.keys(usage).length > 0 ? usage : undefined;
}

/** v2 requests carry `messages`; v1 carries `message` (+ optional `chatHistory`, folded in). */
function extractInput(params: any): unknown {
  if (params?.messages) return params.messages;
  if (Array.isArray(params?.chatHistory) && params.chatHistory.length > 0) {
    return [...params.chatHistory, { role: "USER", message: params?.message }];
  }
  return params?.message;
}

/** Exclusion-list model parameters — everything except the model/input fields of either shape. */
function extractModelParameters(params: any): Record<string, unknown> {
  const { model, message, messages, chatHistory, ...modelParameters } = params ?? {};
  return modelParameters;
}

/** v2 responses carry `message` + `usage.tokens`; v1 carry `text` + `meta.tokens`. */
function extractOutput(response: any): unknown {
  return response?.message ?? response?.text ?? response;
}

function extractUsage(response: any): Record<string, number> | undefined {
  return mapUsage(response?.usage?.tokens ?? response?.meta?.tokens);
}

/** Merge streamed events of either API generation into `{ output, usage }`. v1 events are
 * `eventType`-discriminated (`text-generation` text deltas, `stream-end` with the full response);
 * v2 events are `type`-discriminated (`content-delta` at `delta.message.content.text`,
 * `message-end` with `delta.usage`). The output mirrors the matching non-streaming shape. */
function createCohereStreamAccumulator() {
  let text = "";
  let v1Final: any;
  let v2Usage: any;
  let sawV2 = false;
  return {
    add(event: any) {
      const kind = event?.eventType ?? event?.type;
      if (kind === "text-generation" && typeof event?.text === "string") {
        text += event.text;
      } else if (kind === "stream-end") {
        v1Final = event?.response;
      } else if (kind === "content-delta") {
        sawV2 = true;
        const t = event?.delta?.message?.content?.text;
        if (typeof t === "string") text += t;
      } else if (kind === "message-end") {
        sawV2 = true;
        if (event?.delta?.usage) v2Usage = event.delta.usage;
      }
    },
    finalize() {
      if (sawV2) {
        return {
          output: { role: "assistant", content: [{ type: "text", text }] },
          usage: mapUsage(v2Usage?.tokens),
        };
      }
      return { output: v1Final?.text ?? text, usage: mapUsage(v1Final?.meta?.tokens) };
    },
  };
}

function instrument(
  target: any,
  original: (...args: any[]) => any,
  memoturn: Memoturn,
  options: WrapOptions,
  name: string,
  streaming: boolean,
) {
  return async function call(params: any, ...rest: any[]) {
    const trace = options.trace ?? memoturn.trace({ name: options.traceName ?? "cohere.chat" });
    const generation = trace.generation({
      name,
      model: params?.model,
      provider: "cohere",
      modelParameters: extractModelParameters(params),
      input: extractInput(params),
    });
    try {
      const response = await original.call(target, params, ...rest);
      if (!streaming) {
        generation.end({ output: extractOutput(response), usage: extractUsage(response) });
        return response;
      }
      const accumulator = createCohereStreamAccumulator();
      return tapStream(
        response,
        {
          onChunk: (event: any) => accumulator.add(event),
          onDone: (err, reason) => {
            if (reason === "error") {
              generation.end({ level: "ERROR", statusMessage: String(err), ...accumulator.finalize() });
            } else if (reason === "abandoned") {
              generation.end({
                level: "WARNING",
                statusMessage: "stream ended before completion",
                ...accumulator.finalize(),
              });
            } else {
              generation.end({ ...accumulator.finalize() });
            }
          },
        },
        { idleTimeoutMs: options.streamTimeoutMs },
      );
    } catch (err) {
      generation.end({ level: "ERROR", statusMessage: String(err) });
      throw err;
    }
  };
}

function wrapChatSurface(target: any, memoturn: Memoturn, options: WrapOptions, namePrefix: string): any {
  return new Proxy(target, {
    get(t, prop, receiver) {
      const value = Reflect.get(t, prop, receiver);
      if (prop === "chat" && typeof value === "function") {
        return instrument(t, value, memoturn, options, `${namePrefix}.chat`, false);
      }
      if (prop === "chatStream" && typeof value === "function") {
        return instrument(t, value, memoturn, options, `${namePrefix}.chat`, true);
      }
      if (prop === "v2" && namePrefix === "cohere" && value && typeof value === "object") {
        return wrapChatSurface(value, memoturn, options, "cohere.v2");
      }
      return value;
    },
  });
}

export function wrapCohere<T extends object>(client: T, memoturn: Memoturn, options: WrapOptions = {}): T {
  return wrapChatSurface(client, memoturn, options, "cohere");
}
